powerauth-nextstep/src/main/java/io/getlime/security/powerauth/app/nextstep/service/CredentialCounterService.java

```diff
@@ -137,14 +137,26 @@ public void updateCredentialCounter(CredentialEntity credential, AuthenticationR
     @Transactional
     public ResetCountersResponse resetCounters(ResetCountersRequest request) {
         final List<CredentialEntity> blockedCredentials = credentialRepository.findAllByStatus(CredentialStatus.BLOCKED_TEMPORARY);
+        int resetCounter = 0;
+        // Reset soft counters for credentials with status BLOCKED_TEMPORARY and set status to ACTIVE
         for (CredentialEntity credential: blockedCredentials) {
             credential.setStatus(CredentialStatus.ACTIVE);
             credential.setFailedAttemptCounterSoft(0);
             credential.setTimestampBlocked(null);
+            resetCounter++;
         }
         credentialRepository.saveAll(blockedCredentials);
+        // Reset soft counters for credentials with status ACTIVE with non-zero soft counter
+        final List<CredentialEntity> activeCredentials = credentialRepository.findAllByStatus(CredentialStatus.ACTIVE);
```

Comment:
This is a horrible query that will basically return all customers at once. We should:

- scope it only to those credentials that have non-zero counter
- use paging / limit results (unless Hibernate does some serious magic)

Reply:
I think we can avoid performance issues by executing SQL updates directly. In this case the logic is quite basic, so we don't need to query the data. Thanks for pointing it out.